regulation-worker/internal/delete/api/api.go

```diff
@@ -82,6 +119,39 @@ func (api *APIManager) Delete(ctx context.Context, job model.Job, destConfig map
 		pkgLogger.Errorf("error while decoding response body: %v", err)
 		return model.JobStatusFailed
 	}
+	// Refresh Flow Start
+	var isRefresh bool
+	for _, jobResponse := range jobResp {
+		isRefresh = strings.TrimSpace(jobResponse.AuthErrorCategory) != "" && jobResponse.AuthErrorCategory == oauth.REFRESH_TOKEN
+		if isRefresh {
+			break
+		}
+	}
+	if isRefresh {
+		pkgLogger.Debugf("Refresh flow triggered for %v\n", destName)
+		// Refresh OAuth flow
+		var refSecret *oauth.AuthResponse
+		var errCatStatusCode int
+		refTokenParams := &oauth.RefreshTokenParams{
+			Secret:          accountSecretInfo.Account.Secret,
+			WorkspaceId:     job.WorkspaceID,
+			AccountId:       rudderUserDeleteAccountId.(string),
+			DestDefName:     destName,
+			EventNamePrefix: "refresh_token",
+		}
+		errCatStatusCode, refSecret = api.OAuth.RefreshToken(refTokenParams)
```

Comment:
Just a naive question.
Would it be beneficial if we invalidate/delete the current token we have rather than generate a new one upon failure? So, the next token fetch call would force the module to generate/refresh a new one.

We never know when the next request would come (immediately or 5 mins later), and if the token is generated way early, all that time on the token is expired, doing nothing.

Reply:
### Heads-up about each of the methods

- `FetchToken` only fetches what is present in the config-be DB
- `RefreshToken`
  - if the token sent(from `RefreshToken` method) matches the one present already in config-be DB, refreshes token present in the config-be DB
  - if the token sent(from `RefreshToken` method) doesn’t match the one present already in config-be DB, the token present in the config-be DB is sent
  - if the token sent(from `RefreshToken` method) is not defined or empty or null, the token present in the config-be DB is sent

So if we invalidate or delete the token on failure, the `FetchToken` call wouldn’t work & provide the expired token always(as that is present in DB)

### What happens if we replace `FetchToken` with `RefreshToken` ?

As you see in `router.go`, we send secret information after receive response from transformer proxy(`destinationJob.JobMetadataArray[0].Secret`)
- Based on this, we’d check if the token is present in server-cache is same as the one we used, then try to send this to config-be for refreshing(You can refer [this block of code](https://github.com/rudderlabs/rudder-server/blob/master/router/oauthResponseHandler/oauthResponseHandler.go#L180-L189) to understand)
-  In config-be, we check if the one present in DB is same as the one that we sent
    - If both are same, we refresh the flow, as we know that we have to refresh
    - If both are not same, the one present in DB is sent back(as valid token)

Basically, since at the start, we wouldn’t have any information about the token, the one present in cache is always sent or the one present in config-be DB is sent

It’d not be of much use for us!

Reply:
Let me know if it makes sense

Reply:
My suggestion was to change the OAuth flow to something like this:

If a token is expired, rather than refreshing the token, invalidate it (or delete it). So, the next token fetch request is forced to refresh it.

So, in terms of APIs, we have these two:
- `fetchToken`
  - Fetches the token from cache or DB.
  - If needed, refreshes the token and cache it for future use.
- `invalidateToken`
  - Deletes from cache and DB.

@rudderlabs/integrations-platform would love to hear your thoughts as well!

Reply:
Moving this discussion to an internal forum.